src/ioctl/mod.rs

```diff
@@ -0,0 +1,301 @@
+}
+
+unsafe fn _ioctl(
+    fd: BorrowedFd<'_>,
+    request: RawOpcode,
+    arg: *mut c::c_void,
+) -> Result<IoctlOutput> {
+    crate::backend::io::syscalls::ioctl(fd, request, arg)
+}
+
+unsafe fn _ioctl_readonly(
+    fd: BorrowedFd<'_>,
+    request: RawOpcode,
+    arg: *mut c::c_void,
+) -> Result<IoctlOutput> {
+    crate::backend::io::syscalls::ioctl_readonly(fd, request, arg)
+}
+
+/// A trait defining the properties of an `ioctl` command.
+///
+/// Objects implementing this trait can be passed to [`ioctl`] to make an `ioctl` call. The contents
+/// of the object represent the inputs to the `ioctl` call. The inputs must be convertible to a
+/// pointer through the `as_ptr` method. In most cases, this involves either casting a number to
+/// a pointer, or creating a pointer to the actual data. The latter case is necessary for `ioctl`
+/// calls that modify userland data.
+///
+/// # Safety
+///
+/// This trait is unsafe to implement because it is impossible to guarantee that the `ioctl` call
+/// is safe. The `ioctl` call may be proprietary, or it may be unsafe to call in certain
+/// circumstances.
+///
+/// By implementing this trait, you guarantee that:
+///
+/// - The `ioctl` call expects the input provided by `as_ptr` and produces the output as indicated
+///   by `output`.
+/// - That `output_from_ptr` can safely take the pointer from `as_ptr` and cast it to the correct
+///   type, *only* after the `ioctl` call.
+/// - That `OPCODE` uniquely identifies the `ioctl` call.
+/// - That, for whatever platforms you are targeting, the `ioctl` call is safe to make.
+/// - If `IS_MUTATING` is false, that no userland data will be modified by the `ioctl` call.
+pub unsafe trait Ioctl {
+    /// The type of the output data.
+    ///
+    /// Given a pointer, one should be able to construct an instance of this type.
+    type Output;
+
+    /// The opcode used by this `ioctl` command.
+    ///
+    /// There are different types of opcode depending on the operation. See documentation for the
+    /// [`Opcode`] struct for more information.
+    const OPCODE: Opcode;
+
+    /// Does the `ioctl` mutate any data in the userspace?
+    ///
+    /// If the `ioctl` call does not mutate any data in the userspace, then making this `false`
+    /// enables optimizations that can make the call faster. When in doubt, set this to `true`.
+    ///
+    /// # SAFETY
+    ///
+    /// This should only be set to `false` if the `ioctl` call does not mutate any data in the
+    /// userspace. Undefined behavior may occur if this is set to `false` when it should be `true`.
+    const IS_MUTATING: bool;
+
+    /// Get a pointer to the data to be passed to the `ioctl` command.
+    ///
+    /// See trait-level documentation for more information.
+    fn as_ptr(&mut self) -> *mut c::c_void;
+
+    /// Cast the output data to the correct type.
+    ///
+    /// # Safety
+    ///
+    /// The `ptr` must be the resulting value after a successful `ioctl` call.
```

Comment:
Could you add a comment here saying that `out` is the return value of the `ioctl` call if it doesn't fail, and `ptr` is the pointer that was passed to the `ioctl` call?

And perhaps also rename it to something like `extract_output` or so, because it isn't always using the ptr.